ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockGossipValidator.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Consensys Software Inc., 2022
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.validation;
+
+import static tech.pegasys.teku.infrastructure.async.SafeFuture.completedFuture;
+import static tech.pegasys.teku.spec.config.Constants.VALID_BLOCK_SET_SIZE;
+import static tech.pegasys.teku.statetransition.validation.InternalValidationResult.ignore;
+import static tech.pegasys.teku.statetransition.validation.InternalValidationResult.reject;
+
+import com.google.common.base.Objects;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.collections.LimitedSet;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.constants.Domain;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.execution.ExecutionPayload;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.logic.common.helpers.MiscHelpers;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class BlockGossipValidator {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Spec spec;
+  private final RecentChainData recentChainData;
+  private final GossipValidationHelper gossipValidationHelper;
+  private final Set<SlotAndProposer> receivedValidBlockInfoSet =
+      LimitedSet.createSynchronized(VALID_BLOCK_SET_SIZE);
+
+  public BlockGossipValidator(
+      final Spec spec,
+      final RecentChainData recentChainData,
+      final GossipValidationHelper gossipValidationHelper) {
+    this.spec = spec;
+    this.recentChainData = recentChainData;
+    this.gossipValidationHelper = gossipValidationHelper;
+  }
+
+  SafeFuture<InternalValidationResult> validate(
+      final SignedBeaconBlock block, final boolean isLocal) {
```

Comment:
I think `isLocal` could be a bit misleading here, because we don't pass standard locally produced blocks here, maybe we could somehow highlight it in the name of this flag. I don't know, `isLocalUntrusted`?

Reply:
But from the other side, BlockGossipValidator shouldn't know too much over his job. So maybe it's ok to leave it as is. Just want to remember what's happening there after 2 months.

Reply:
I don't get `isLocalUntrusted` but maybe `isLocallyProduced`?